core/include/gnuradio-4.0/HistoryBuffer.hpp

```diff
@@ -100,23 +113,176 @@ class HistoryBuffer {
     /**
      * @brief Adds a range of elements the end expiring the oldest elements beyond the buffer's capacities.
      */
-    template<typename Iter>
-    constexpr void push_back_bulk(Iter cbegin, Iter cend) noexcept {
+    template<std::input_iterator Iter>
+    constexpr void push_back(Iter cbegin, Iter cend) noexcept {
         const auto nSamplesToCopy = std::distance(cbegin, cend);
         Iter       optimizedBegin = static_cast<std::size_t>(nSamplesToCopy) > _capacity ? std::prev(cend, static_cast<std::ptrdiff_t>(_capacity)) : cbegin;
         for (auto it = optimizedBegin; it != cend; ++it) {
             push_back(*it);
         }
     }
 
+    template<std::input_iterator Iter>
+    constexpr void push_back_bulk(Iter cbegin, Iter cend) noexcept {
+        return push_back(cbegin, cend);
+    }
+
     /**
      * @brief Adds a range of elements the end expiring the oldest elements beyond the buffer's capacities.
      */
-    template<typename Range>
+    template<std::ranges::range Range>
+    constexpr void push_back(const Range& range) noexcept {
+        push_back_bulk(range.cbegin(), range.cend());
+    }
+
+    template<std::ranges::range Range>
     constexpr void push_back_bulk(const Range& range) noexcept {
         push_back_bulk(range.cbegin(), range.cend());
     }
 
+    /**
+     * @brief Adds an element to the front expiring the oldest element beyond the buffer's capacities.
+     */
+    constexpr void push_front(const T& value) noexcept {
+        if (_size == _capacity) {
+            if (++_write_position == _capacity) {
+                _write_position = 0U;
+            }
+        } else {
+            ++_size;
+        }
+
+        std::size_t insertPos = _write_position + (_size - 1);
+        if (insertPos >= _capacity) {
+            insertPos -= _capacity;
+        }
+
+        _buffer[insertPos]             = value;
+        _buffer[insertPos + _capacity] = value;
+    }
+
+    /**
+     * @brief Inserts a contiguous range [first, last) so that the new items become
+     *        the "newest" in the ring, with operator[](0) always the oldest.
+     */
+    template<std::input_iterator Iter>
+    constexpr void push_front(Iter first, Iter last) noexcept {
```

Comment:
Following the previous comment, this would be `prepend_range`.

Reply:
I went for the [std::deque<T>](https://en.cppreference.com/w/cpp/container/deque) API syntax with some sprinkles of `std::vector<T>`. Hope this make sense. Didn't want to break too much of the existing API/use elsewhere.

:-)

Reply:
`std::deque` also has single-value push_back/push_front and append_range/prepend_range for multiple values.